### PR TITLE
Feature/cvsb-3331

### PR DESCRIPTION
**Pull request by arupdvsa**
- moved codebase to typescript (Initial draft)
- lamda-tester moved to Integration test section from Unit tests
- fragmented Unit Tests to reduce code cyclomatics { too many tests and context logic in a single file }

### CODE REVIEW

File: src/models/HTTPResponse.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  /**
+   * Constructor for the HTTPResponse class
+   * @param statusCode the HTTP status code
+   * @param body - the response body
+   * @param headers - optional - the response headers
+   */
+  constructor(statusCode: number, body: any, headers = {}) {
+    super();
+
+    if (headers) { this.headers = headers; }
```
Comment (KHodderDVSA) [inline src/models/HTTPResponse.ts]: I know this is everywhere, but just realised - does this need the "if" if headers defaults to an empty object?

File: src/models/HTTPResponse.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import {APIGatewayProxyResult} from "aws-lambda";
+export class HTTPResponse extends Error implements APIGatewayProxyResult {
```
Comment (KHodderDVSA) [inline src/models/HTTPResponse.ts]: Should this be an error? I feel like that would have bad consequences. May have been an error in my 3330 code

File: src/utils/Configuration.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (KHodderDVSA) [inline src/utils/Configuration.ts]: The other services have used a slightly different model for config - using a config.yml, and Configuration.ts just parses and splits up the pieces of config. Not sure if you have a strong preference, but I was inclined to go for consistency.

File: tests/integration/defects.intTest.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+describe('defects', () => {
+    describe('getDefects', () => {
+        context('when database is populated', () => {
+            let defectsService: DefectsService;
+            const defectsData = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../resources/defects.json'), 'utf8'));
+            let defectsDAO = null;
+
+            before((done) => {
+                defectsService = Injector.resolve<DefectsService>(DefectsService, [DefectsDAO]);
```
Comment (KHodderDVSA) [inline tests/integration/defects.intTest.ts]: Real DAO, so matches expected type - no need for injection here.

File: tests/integration/defects.intTest.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import fs from "fs";
+import path from "path";
+import { DefectsService } from '../../src/services/DefectsService';
+import { DefectsDAO } from '../../src/models/DefectsDAO';
+import { Injector } from "../../src/models/injector/Injector";
+
+const url = "http://localhost:3004/";
+const request = supertest(url)
+
+describe('defects', () => {
```
Comment (KHodderDVSA) [inline tests/integration/defects.intTest.ts]: Should be "Defects service" - I know it's a tie over from before, but may as well fix it.

File: tests/integration/defects.intTest.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    describe('getDefects', () => {
+        context('when database is populated', () => {
+            let defectsService: DefectsService;
+            const defectsData = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../resources/defects.json'), 'utf8'));
+            let defectsDAO = null;
```
Comment (KHodderDVSA) [inline tests/integration/defects.intTest.ts]: unused

File: src/models/HTTPResponse.ts (hunk 2 above)
Comment (KHodderDVSA) [reply src/models/HTTPResponse.ts]: hmmm... test number does this. Still feels wrong.

File: src/models/DefectsDAO.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        //         {
+        //             PutRequest:
+        //             {
+        //                 Item: defectItem
+        //             }
+        //         })
+        // })
+
+        
+        defectItems.map((defectItem: any) => {
```
[the remark this answers is not in the document]
Comment (arupdvsa) [reply src/models/DefectsDAO.ts]: makes sense I forgot

File: src/models/HTTPResponse.ts (hunk 1 above)
Comment (arupdvsa) [reply src/models/HTTPResponse.ts]: good point ... slipped my eyes ... `if {header)` will always evaluate to truthy ... due to default ... hence the `if` check is unnecessary ... the reason for default empty object is understandable

File: src/utils/Configuration.ts (hunk 3 above)
Comment (arupdvsa) [reply src/utils/Configuration.ts]: The idea of config as yml, as used in3330 is definitely better than the approach used in defects.  Additionally there are also models which I found are just code repetitions across the lambdas, contrary to the DRY principles.
Probably as next stage we can have a common npm repo which will contain these common features/functionalities and can be directly sourced across the lambdas so we can facilitate the DRY principles .. 

On the same note I was thinking of the pattern used in jest for mocking dynamodb endpoint as facilitated by jest framework.

For these reasons I've not tried to optimise these common features as part of this commit, but to use existing and await for an agreed pattern which leverages the agreed ideas and principles across all lambdas

File: src/models/HTTPResponse.ts (hunk 2 above)
Comment (arupdvsa) [reply src/models/HTTPResponse.ts]: good point ... I've made the changes

File: tests/integration/defects.intTest.ts (hunk 4 above)
Comment (arupdvsa) [reply tests/integration/defects.intTest.ts]: agreed

File: tests/integration/defects.intTest.ts (hunk 5 above)
Comment (arupdvsa) [reply tests/integration/defects.intTest.ts]: yeah

File: tests/integration/defects.intTest.ts (hunk 6 above)
Comment (arupdvsa) [reply tests/integration/defects.intTest.ts]: now used

**Comment by KHodderDVSA** [comment]
Closing, and moving merge to 3330, s that is where testing will happen, and any subsequent changes

**Pull request closed without merging**